pystac_client/collection_search.py

```diff
@@ -0,0 +1,505 @@
+import warnings
+from datetime import datetime, timezone
+from functools import lru_cache
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Dict,
+    Iterator,
+    List,
+    Optional,
+    Sequence,
+)
+
+from pystac import Collection
+
+from pystac_client._utils import Modifiable, call_modifier
+from pystac_client.conformance import ConformanceClasses
+from pystac_client.free_text import sqlite_text_search
+from pystac_client.search import (
+    BaseSearch,
+    BBox,
+    BBoxLike,
+    Datetime,
+    DatetimeLike,
+    FieldsLike,
+    FilterLangLike,
+    FilterLike,
+    QueryLike,
+    SortbyLike,
+)
+from pystac_client.stac_api_io import StacApiIO
+from pystac_client.warnings import DoesNotConformTo
+
+if TYPE_CHECKING:
+    from pystac_client import client as _client
+
+
+TemporalInterval = Sequence[Optional[datetime]]
```

Comment:
I don't know if this breaks stuff below, but since we know (and expect in code) that this is always two items, IMO we should specify it in the type.

```suggestion
TemporalInterval = Tuple[[Optional[datetime], Optional[datetime]]]
```